File: engine/src/solver/buckling.rs
use crate::types::*;
use crate::linalg::*;
use super::dof::DofNumbering;
use super::assembly::*;
use super::geometric_stiffness::build_kg_from_forces_2d;

/// Buckling analysis result.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BucklingResult {
    pub modes: Vec<BucklingMode>,
    pub n_dof: usize,
    pub element_data: Vec<ElementBucklingData>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BucklingMode {
    pub load_factor: f64,
    pub displacements: Vec<Displacement>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ElementBucklingData {
    pub element_id: usize,
    pub axial_force: f64,
    pub critical_force: f64,
    pub k_effective: f64,
    pub effective_length: f64,
    pub length: f64,
    pub slenderness: f64,
}

/// Solve 2D buckling analysis.
/// Solves K·φ = λ·(-Kg)·φ where Kg is from linear axial forces.
pub fn solve_buckling_2d(
    input: &SolverInput,
    num_modes: usize,
) -> Result<BucklingResult, String> {
    // 1. Linear solve to get axial forces
    let linear = super::linear::solve_2d(input)?;
    let dof_num = DofNumbering::build_2d(input);
    let nf = dof_num.n_free;
    let n = dof_num.n_total;

    if nf == 0 {
        return Err("No free DOFs".into());
    }

    // 2. Build geometric stiffness from linear axial forces
    let kg_full = build_kg_from_forces_2d(input, &dof_num, &linear.element_forces);

    // 3. Extract free-DOF submatrices
    let free_idx: Vec<usize> = (0..nf).collect();
    let asm = assemble_2d(input, &dof_num);
    let k_ff = extract_submatrix(&asm.k, n, &free_idx, &free_idx);

    // Negate Kg (we solve K·φ = λ·(-Kg)·φ for positive eigenvalues)
    let kg_ff_raw = extract_submatrix(&kg_full, n, &free_idx, &free_idx);
    let mut neg_kg_ff = vec![0.0; nf * nf];
    for i in 0..nf * nf {
        neg_kg_ff[i] = -kg_ff_raw[i];
    }

    // Check if any element is in compression
    let has_compression = linear.element_forces.iter().any(|ef| {
        (ef.n_start + ef.n_end) / 2.0 < -1e-6
    });
    if !has_compression {
        return Err("No compressed elements — buckling not applicable".into());
    }

    // 4. Solve generalized eigenvalue: (-Kg)·φ = μ·K·φ  (K is SPD)
    //    Then λ = 1/μ are the critical load factors.
    //    We swap A=-Kg, B=K because K is SPD (required for Cholesky).
    //    Use dense solver: buckling needs all eigenvalues to find positive μ values.
    let result = solve_generalized_eigen(&neg_kg_ff, &k_ff, nf, 200)
        .ok_or_else(|| "Eigenvalue decomposition failed — stiffness matrix issue".to_string())?;

    // 5. Extract positive μ values → λ = 1/μ (critical load factors)
    //    μ are sorted ascending; we want positive μ → smallest λ = 1/largest_μ.
    let num_modes = num_modes.min(nf);
    let mut mode_pairs: Vec<(f64, usize)> = Vec::new();
    for (idx, &mu) in result.values.iter().enumerate() {
        if mu > 1e-12 {
            let lambda = 1.0 / mu;
            mode_pairs.push((lambda, idx));
        }
    }
    // Sort by ascending load factor (smallest λ first)
    mode_pairs.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());

    let mut modes = Vec::new();
    for &(lambda, idx) in mode_pairs.iter().take(num_modes) {
        let mut u_mode = vec![0.0; n];
        let mut max_disp = 0.0f64;
        for i in 0..nf {
            u_mode[i] = result.vectors[i * nf + idx];
            max_disp = max_disp.max(u_mode[i].abs());
        }
        if max_disp > 1e-20 {
            for i in 0..nf {
                u_mode[i] /= max_disp;
            }
        }

        let displacements = super::linear::build_displacements_2d(&dof_num, &u_mode);
        modes.push(BucklingMode {
            load_factor: lambda,
            displacements,
        });
    }

    if modes.is_empty() {
        return Err("No positive buckling load factors found".into());
    }

    // 6. Per-element buckling data
    let lambda_cr = modes[0].load_factor;
    let mut element_data = Vec::new();
    for ef in &linear.element_forces {
        let n_avg = (ef.n_start + ef.n_end) / 2.0;
        if n_avg >= -1e-6 {
            continue; // Skip tension elements
        }
        let elem = input.elements.values().find(|e| e.id == ef.element_id).unwrap();
        let sec = input.sections.values().find(|s| s.id == elem.section_id).unwrap();
        let l = ef.length;
        let pcr = lambda_cr * n_avg.abs();
        let r = if sec.a > 1e-20 { (sec.iz / sec.a).sqrt() } else { 0.0 };
        let k_eff = if pcr > 1e-6 && r > 1e-12 {
            let le = std::f64::consts::PI * (input.materials.values()
                .find(|m| m.id == elem.material_id).unwrap().e * 1000.0
                * sec.iz / pcr).sqrt();
            le / l
        } else {
            1.0
        };
        let le = k_eff * l;
        let slenderness = if r > 1e-12 { le / r } else { 0.0 };

        element_data.push(ElementBucklingData {
            element_id: ef.element_id,
            axial_force: n_avg,
            critical_force: pcr,
            k_effective: k_eff,
            effective_length: le,
            length: l,
            slenderness,
        });
    }

    Ok(BucklingResult {
        modes,
        n_dof: nf,
        element_data,
    })
}

File: web/src/lib/engine/buckling.ts
// Linear buckling analysis (Euler buckling)
// Solves: K·φ = λ·(-Kg)·φ  where λ_cr = critical load factor

import type { SolverInput } from './types';
import { buildDofNumbering, assemble, solveLU, computeInternalForces, nodeDistance } from './solver-js';
import { assembleKg } from './geometric-stiffness';
import { solveGeneralizedEigen, choleskySolve } from './matrix-utils';
import { t } from '../i18n';

export interface BucklingMode {
  /** Critical load factor (multiply applied loads by this to get buckling) */
  loadFactor: number;
  /** Mode shape displacements per node */
  displacements: Array<{ nodeId: number; ux: number; uy: number; rz: number }>;
}

/** Per-element buckling data */
export interface ElementBucklingData {
  elementId: number;
  /** Axial force from linear analysis (kN). Negative = compression. */
  axialForce: number;
  /** Critical axial force for this element: Pcr = λ_cr × |N| (kN) */
  criticalForce: number;
  /** Effective length factor K = Le/L. Derived from Pcr = π²EI/(KL)². */
  kEffective: number;
  /** Effective length Le = K × L (m) */
  effectiveLength: number;
  /** Element length (m) */
  length: number;
  /** Slenderness ratio λ = KL/r where r = √(I/A) */
  slenderness: number;
}

export interface BucklingResult {
  modes: BucklingMode[];
  nDof: number;
  /** Per-element buckling data (only for compressed elements) */
  elementData: ElementBucklingData[];
}

/**
 * Linear buckling analysis.
 * 1. Solve linear → get axial forces N
 * 2. Build Kg from N
 * 3. Solve eigenvalue: K·φ = λ·(-Kg)·φ
 * 4. λ_cr = smallest positive eigenvalue
 */
export function solveBuckling(
  input: SolverInput,
  numModes?: number,
): BucklingResult | string {
  const dofNum = buildDofNumbering(input);
  const nf = dofNum.nFree;
  const nt = dofNum.nTotal;

  if (nf === 0) return t('buckling.noFreeDofs');
  if (nf > 500) return t('buckling.modelTooLarge');

  // Step 1: Linear solve
  const { K, F } = assemble(input, dofNum);

  const Kff = new Float64Array(nf * nf);
  for (let i = 0; i < nf; i++) {
    for (let j = 0; j < nf; j++) {
      Kff[i * nf + j] = K[i * nt + j];
    }
  }

  const Ff = new Float64Array(nf);
  for (let i = 0; i < nf; i++) Ff[i] = F[i];

  let uFree: Float64Array;
  try {
    uFree = choleskySolve(new Float64Array(Kff), new Float64Array(Ff), nf)
          ?? solveLU(new Float64Array(Kff), new Float64Array(Ff), nf);
  } catch {
    return t('buckling.linearSolveError');
  }

  // Build full displacement vector
  const uAll = new Float64Array(nt);
  for (let i = 0; i < nf; i++) uAll[i] = uFree[i];

  // Step 2: Get axial forces and build Kg
  const elementForces = computeInternalForces(input, dofNum, uAll);
  const Kg = assembleKg(input, dofNum, elementForces);

  // Check Kg is not zero
  let kgNorm = 0;
  for (let i = 0; i < nf * nf; i++) kgNorm += Kg[i] * Kg[i];
  if (kgNorm < 1e-20) return t('buckling.noCompression');

  // Step 3: Generalized eigenvalue problem
  // K·φ = λ·(-Kg)·φ  →  (-Kg)·φ = μ·K·φ  where μ = 1/λ
  // Use Cholesky-based solver with K as the SPD matrix (B).
  // This correctly preserves symmetry via the L⁻¹·(-Kg)·L⁻ᵀ transformation.
  const negKg = new Float64Array(nf * nf);
  for (let i = 0; i < nf * nf; i++) negKg[i] = -Kg[i];

  const eigen = solveGeneralizedEigen(negKg, Kff, nf);
  if (!eigen) return t('buckling.choleskyError');

  // eigenvalues μ satisfy (-Kg)·φ = μ·K·φ → K·φ = (1/μ)·(-Kg)·φ → λ = 1/μ
  // We want positive λ (from positive μ): smallest λ = largest positive μ
  const nModes = Math.min(numModes ?? 4, nf);
  const modes: BucklingMode[] = [];

  // Collect positive eigenvalues → λ_cr = 1/μ, sort by λ_cr ascending
  const candidates: Array<{ lambdaCr: number; modeIdx: number }> = [];
  for (let i = 0; i < nf; i++) {
    const mu = eigen.values[i];
    if (mu > 1e-10) {
      candidates.push({ lambdaCr: 1 / mu, modeIdx: i });
    }
  }
  candidates.sort((a, b) => a.lambdaCr - b.lambdaCr);

  for (let m = 0; m < Math.min(nModes, candidates.length); m++) {
    const { lambdaCr, modeIdx } = candidates[m];

    // Extract mode shape
    const phi = new Float64Array(nf);
    for (let i = 0; i < nf; i++) {
      phi[i] = eigen.vectors[i * nf + modeIdx];
    }

    // Normalize
    let maxAbs = 0;
    for (let i = 0; i < nf; i++) {
      if (Math.abs(phi[i]) > maxAbs) maxAbs = Math.abs(phi[i]);
    }
    if (maxAbs > 0) {
      for (let i = 0; i < nf; i++) phi[i] /= maxAbs;
    }

    // Map to node displacements
    const displacements: BucklingMode['displacements'] = [];
    for (const [nodeId] of input.nodes) {
      const ux = getPhiVal(dofNum, phi, nodeId, 0, nf);
      const uy = getPhiVal(dofNum, phi, nodeId, 1, nf);
      const rz = dofNum.dofsPerNode > 2 ? getPhiVal(dofNum, phi, nodeId, 2, nf) : 0;
      displacements.push({ nodeId, ux, uy, rz });
    }

    modes.push({ loadFactor: lambdaCr, displacements });
  }

  if (modes.length === 0) return t('buckling.noModesFound');

  // Compute per-element buckling data using the first (critical) mode
  const lambdaCr1 = modes[0].loadFactor;
  const elementData: ElementBucklingData[] = [];

  for (const ef of elementForces) {
    const elem = input.elements.get(ef.elementId);
    if (!elem) continue;
    const nodeI = input.nodes.get(elem.nodeI);
    const nodeJ = input.nodes.get(elem.nodeJ);
    if (!nodeI || !nodeJ) continue;

    const N = (ef.nStart + ef.nEnd) / 2; // average axial force
    if (N >= -1e-10) continue; // only compressed elements (N < 0 = compression)

    const L = nodeDistance(nodeI, nodeJ);
    if (L < 1e-12) continue;

    const sec = input.sections.get(elem.sectionId);
    const mat = input.materials.get(elem.materialId);
    if (!sec || !mat) continue;

    const EI = mat.e * 1000 * sec.iz; // kN·m² (e in MPa → ×1000 for kPa)
    const absN = Math.abs(N);
    const Pcr = lambdaCr1 * absN; // critical force for this element

    // Keff from Pcr = π²EI/(Keff·L)²
    // Keff = π·√(EI/Pcr) / L
    const kEff = Pcr > 1e-15 ? Math.PI * Math.sqrt(EI / Pcr) / L : Infinity;

    // Effective length
    const Le = kEff * L;

    // Radius of gyration r = √(Iz/A)
    const r = Math.sqrt(sec.iz / sec.a);
    const slenderness = Le / r;

    elementData.push({
      elementId: ef.elementId,
      axialForce: N,
      criticalForce: Pcr,
      kEffective: kEff,
      effectiveLength: Le,
      length: L,
      slenderness,
    });
  }

  return { modes, nDof: nf, elementData };
}

function getPhiVal(
  dofNum: { map: Map<string, number>; dofsPerNode: number },
  phi: Float64Array,
  nodeId: number,
  localDof: number,
  nFree: number,
): number {
  if (localDof >= dofNum.dofsPerNode) return 0;
  const key = `${nodeId}:${localDof}`;
  const idx = dofNum.map.get(key);
  if (idx === undefined || idx >= nFree) return 0;
  return phi[idx];
}
